src/transport/pci.rs

```diff
@@ -366,6 +366,18 @@ impl Transport for PciTransport {
     }
 }
 
+impl PciTransport {
+    /// Get the MSI-X vector for queue.
+    pub fn get_queue_msix_vector(&self) -> u16 {
+        field_shared!(self.common_cfg, queue_msix_vector).read()
+    }
+
+    /// Set the MSI-X vector for queue.
```

Comment:
Before setting MSI-X vectors, the driver will need to check whether the device has the MSI-X capability, and check its MSI-X table size. I guess you either need separate methods for this on the transport, or some method to handle MSI-X setup generally.

I guess the `ack_interrupt` implementation will also need to change to support MSI-X interrupts when it is being used.

Reply:
yes code has to read the capability and such. i didn't add code for this because the code in my os for this is done outside of the virtio library, as i support non-virtio devices as well.

the interrupt ack register seems to only be used for "legacy" interrupts, according to the virtio spec.

Reply:
The `isr_status` register is only for legacy interrupts, but the `Transport::ack_interrupt` method should work either way. What I'm saying is that `PciTransport` should have some logic for negotiating whether MSI-X is being used or not, and if it is used then the behaviour of `ack_interrupt` should change.